vim-patch:partial:8.2.3578

shadmansaleh opened the pull request:
#### vim-patch:partial:8.2.3578: manipulating highlighting is complicated

Problem:    Manipulating highlighting is complicated.
Solution:   Add the hlget() and hlset() functions. (Yegappan Lakshmanan,
            closes vim/vim#9039)
https://github.com/vim/vim/commit/d1a8d658e1b16cf8579fc72cf7aa6a29a57ff5ef

skipped support for highlight-start, highlight-stop, highlight-ctermul, highlight-term, highlight-font

partial skip src/nvim/syntax.c
partial skip src/nvim/testdir/test_highlight.vim
partial skip runtime/doc/eval.txt

shadmansaleh commented:
Skipped support for highlight-start, highlight-stop, highlight-ctermul, highlight-term, highlight-font since they aren't available in neovim . Should I mark it as partial port ?

Also it seems like docs for highlight-start & stop is included but the features are not 🤷‍♂️

shadmansaleh commented:
> so this may need to be partial'

Ok I've marked it as partial. Do check if this is the right way to do a partial port.

> ctermul needs v8.2.0863 which isn't ported

Did @janlazo hand over his note book , how do you even figure out what patch is what 😆

I did manage to locate v8.2.0863 patch for ctermul by blaming ctermul docs in vim repo. But couldn't get anything on start,stop,font,term .

These lone [docs](https://github.com/neovim/neovim/blob/8cbe100fcc9c5d34b07f417a82b745803fa3ef19/runtime/doc/syntax.txt#L4880-L4897) only added to the confusion .
Should these docs be removed since the features aren't ported yet ?

janlazo commented:
vim-patch.sh doesn't ignores `vim-patch:partial:` so it's techincally fine.
However, it can be confusing because `vim-patch:` should imply full port.

zeertzjq commented:
Maybe this has something to do with licensing?

shadmansaleh commented:
> However, it can be confusing because vim-patch: should imply full port.

I don't remember seeing any guide for partial port . I just copied structure from newest partial patch commit on the log. Which at time was https://github.com/neovim/neovim/commit/8cbe100fcc9c5d34b07f417a82b745803fa3ef19

Do let me know if partial patch is meant to be documented differently. 

> Maybe this has something to do with licensing?

@zeertzjq sorry I have no clue what you meant. Can you please elaborate?

zeertzjq commented:
```
Neovim contributions since b17d96 are licensed under the Apache 2.0 license, except for contributions copied from Vim (identified by the vim-patch token).
```

shadmansaleh commented:
> Neovim contributions since b17d96 are licensed under the Apache 2.0 license, except for contributions copied from Vim (identified by the vim-patch token).

That's general case for all patches right?

janlazo commented on src/nvim/testdir/test_highlight.vim:
```diff
@@ -761,4 +761,137 @@ func Test_highlight_default_colorscheme_restores_links()
+  highlight link MyHLGroup IncSearch
+  call assert_equal('IncSearch', hlget('MyHLGroup')[0].linksto)
+  highlight clear MyHLGroup
+  " call assert_equal([], hlget(v:null))
```
This fails?

shadmansaleh replied:
Yeah it fails with v:null . Didn't know about v:_null_string v:_null_dict . Those will probably fix this . I was wondering how to get a null with string type.

janlazo commented on src/nvim/testdir/test_highlight.vim:
```diff
@@ -761,4 +761,137 @@ func Test_highlight_default_colorscheme_restores_links()
+
+" Test for the hlset() function
+func Test_hlset()
+  " call assert_equal(0, hlset(v:null))
```
`v:_null_list`

janlazo commented on src/nvim/syntax.c:
```diff
@@ -8754,6 +8755,391 @@ RgbValue name_to_color(const char *name)
+  int         i;
+
+  dict = tv_dict_alloc();
+  if (dict == NULL) {
```
tv_dict_alloc() does not return NULL.

shadmansaleh replied:
right 🤦‍♂️

janlazo commented on src/nvim/syntax.c:
```diff
@@ -8754,6 +8755,391 @@ RgbValue name_to_color(const char *name)
+  dict_T      *attr_dict;
+  int         hlgid;
+
+  dict = tv_dict_alloc();
```
See https://github.com/neovim/neovim/pull/16270/files#r752833547

janlazo commented on src/nvim/syntax.c:
```diff
@@ -8754,6 +8755,391 @@ RgbValue name_to_color(const char *name)
+
+  tv_list_alloc_ret(rettv, kListLenMayKnow);
+
+  if (argvars[0].v_type != VAR_UNKNOWN) {
```
Could have ported the equivalent Vim function in eval/typval.c instead of this.

shadmansaleh replied:
Should I put them part of this patch?  In a separate commit? Or as a different vim-patch (Do you know what vim-path I'll need to port. I'm not sure how to find a patch number TBH :| )?

zeertzjq replied:
May be <https://github.com/vim/vim/commit/83494b4ac61898f687d6ef9dce4bad5802fb8e51>

shadmansaleh replied:
Thanks @zeertzjq I'll take a look at that patch.

shadmansaleh replied:
@janlazo @zeertzjq most of the changes related seems to be type checking related to vim9. Aren't those NA? Should I just pick the changes in `src/typeval.c` ?

janlazo replied:
Yes.

shadmansaleh replied:
@janlazo is porting those patches in this pr strictly necessary ? Can this be merged without those and those patches ported on a later pr ?

3N4N commented:
Is this not gonna be picked up?

dundargoc commented:
@shadmansaleh can you please rebase?

shadmansaleh commented:
> @shadmansaleh can you please rebase?

I'll try to take another look at it by the weekend.

shadmansaleh committed:
vim-patch:partial:8.2.3578: manipulating highlighting is complicated

Problem:    Manipulating highlighting is complicated.
Solution:   Add the hlget() and hlset() functions. (Yegappan Lakshmanan,
            closes vim/vim#9039)
https://github.com/vim/vim/commit/d1a8d658e1b16cf8579fc72cf7aa6a29a57ff5ef

skipped support for highlight-start, highlight-stop, highlight-ctermul, highlight-term, highlight-font

partial skip src/nvim/syntax.c
partial skip src/nvim/testdir/test_highlight.vim
partial skip runtime/doc/eval.txt

shadmansaleh committed:
fix linter errors

shadmansaleh committed:
address review comments1

shadmansaleh committed:
fix help docs validate test

shadmansaleh commented:
Ok rebased on master and tests are passing now.

zeertzjq commented on runtime/doc/builtin.txt:
```diff
@@ -3977,6 +3979,88 @@ hlexists({name})					*hlexists()*
+		Example(s): >
+			" add bold attribute to the Visual highlight group
+			:call hlset([#{name: 'Visual',
+					\ term: #{reverse: 1 , bold: 1}}])
```
```suggestion
					\ cterm: #{reverse: 1 , bold: 1}}])
```

zeertzjq commented on src/nvim/syntax.c:
```diff
@@ -18,7 +18,9 @@
 #include "nvim/cursor_shape.h"
 #include "nvim/drawscreen.h"
 #include "nvim/eval.h"
+#include "nvim/eval/typval.h"
 #include "nvim/eval/vars.h"
+#include "nvim/ex_cmds2.h"
```
The two added headers are no longer needed.

zeertzjq commented on src/nvim/syntax.h:
```diff
@@ -4,6 +4,7 @@
 #include <stdbool.h>
 
 #include "nvim/buffer_defs.h"
+#include "nvim/eval/funcs.h"
```
This is no longer needed.

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
   return -1;
 }
 
+// Convert each of the highlight attribute bits (bold, standout, underline,
+// etc.) set in 'hlattr' into a separate boolean item in a Dictionary with
+// the attribute name as the key.
+
```
```suggestion
/// Convert each of the highlight attribute bits (bold, standout, underline,
/// etc.) set in 'hlattr' into a separate boolean item in a Dictionary with
/// the attribute name as the key.
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  return dict;
+}
+
+// Return the attributes of the highlight group at index 'hl_idx' as a
+// Dictionary. If 'resolve_link' is true, then resolves the highlight group
+// links recursively.
+
```
```suggestion
/// Return the attributes of the highlight group at index 'hl_idx' as a
/// Dictionary. If 'resolve_link' is true, then resolves the highlight group
/// links recursively.
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  return NULL;
+}
+
+// "hlget([name])" function
+// Return the attributes of a specific highlight group (if specified) or all
+// the highlight groups.
+
```
```suggestion
/// "hlget([name])" function
/// Return the attributes of a specific highlight group (if specified) or all
/// the highlight groups.
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+
+  tv_list_alloc_ret(rettv, kListLenMayKnow);
+
+  if (argvars[0].v_type != VAR_UNKNOWN) {
+    if (argvars[0].v_type != VAR_STRING) {
+      emsg("E1174: String required for argument 0");
+      return;
+    } else if (argvars[1].v_type != VAR_UNKNOWN
+               && (argvars[1].v_type != VAR_BOOL
+                   && !(argvars[1].v_type == VAR_NUMBER
+                        && (argvars[1].vval.v_number == 0
+                            || argvars[1].vval.v_number == 1)))) {
+      emsg("E1212: Bool required for argument 1");
+      return;
+    }
+  }
```
Port the `tv_check_for_opt_string_arg()` and `tv_check_for_opt_bool_arg()` to `eval/typval.c`

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  }
+}
+
+// Returns the string value at 'dict[key]'. Returns NULL, if 'key' is not in
+// 'dict' or the value is not a string type. If the value is not a string type
+// or is NULL, then 'error' is set to true.
+
```
```suggestion
/// Returns the string value at 'dict[key]'. Returns NULL, if 'key' is not in
/// 'dict' or the value is not a string type. If the value is not a string type
/// or is NULL, then 'error' is set to true.
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  return (char_u *)di->di_tv.vval.v_string;
+}
+
+// Convert the highlight attribute Dictionary at 'dict[key]' into a string
+// value in 'attr_str' of length 'len'. Returns false if 'dict[key]' is not a
+// Dictionary or is NULL.
+
```
```suggestion
/// Convert the highlight attribute Dictionary at 'dict[key]' into a string
/// value in 'attr_str' of length 'len'. Returns false if 'dict[key]' is not a
/// Dictionary or is NULL.
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  return true;
+}
+
+// Add or update a highlight group using 'dict' items. Returns true if
+// successfully updated the highlight group.
+
```
```suggestion
/// Add or update a highlight group using 'dict' items. Returns true if
/// successfully updated the highlight group.
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  return true;
+}
+
+// "hlset([{highlight_attr}])" function
+// Add or modify highlight groups
+
```
```suggestion
/// "hlset([{highlight_attr}])" function
/// Add or modify highlight groups
```

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+    }
+
+    if (argvars[1].v_type != VAR_UNKNOWN) {
+      if (argvars[1].v_type == VAR_BOOL) {
+        resolve_link = argvars[1].vval.v_bool == kBoolVarTrue ? true : false;
+      } else if (argvars[1].v_type == VAR_NUMBER) {
+        resolve_link = argvars[1].vval.v_number == 1 ? true : false;
```
This is obviously wrong as it treats a number larger than 1 as `false`. Not sure what to do here. Maybe just `#define tv_get_bool_chk tv_get_number_chk`?

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+  }
+
+  for (i = 0; i < (int)(sizeof(hl_name_table)/sizeof(hl_name_table[0])); i++) {
+    if (tv_dict_get_number(attrdict, (const char *)hl_name_table[i]) == 1) {
```
This also treats numbers larger than 1 as `false`.

zeertzjq commented:
I've opened a PR to deal with the `get_bool` functions: #20774

shadmansaleh committed:
address review comments 2

shadmansaleh committed:
chore: Port type checker functions from vim

ports:
tv_check_for_opt_string_arg
tv_check_for_bool_arg
tv_check_for_bool_arg
tv_check_for_list_arg

shadmansaleh committed:
address review comments 3

shadmansaleh commented:
@zeertzjq  is it ok to pull the type checker functions directly from vim source without the rest of the patches https://github.com/neovim/neovim/pull/16270/commits/f0eb636d59b247b1121c0ad310fbcd1720dde736 ?

it seems to have been done before.

zeertzjq commented:
Yes this has been done before.

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,376 @@ RgbValue name_to_color(const char *name, int *idx)
+  attrdict = di->di_tv.vval.v_dict;
+
+  // If the attribute dict is empty, then return NONE to clear the attributes
+  if (attrdict->dv_hashtab.ht_used == 0) {
```
`tv_dict_len()`

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,376 @@ RgbValue name_to_color(const char *name, int *idx)
+      return false;
+    }
+
+    vim_snprintf((char *)IObuff, IOSIZE, "link %s %s", name, linksto);
+    do_highlight((char *)IObuff, false, false);
```
Redundant casts

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -2901,6 +2901,376 @@ RgbValue name_to_color(const char *name, int *idx)
+      goto error;
+    }
+  }
+  if (dict->dv_hashtab.ht_used == 2) {
```
`tv_dict_len()`

shadmansaleh committed:
address review comments 4

vimpostor commented:
@zeertzjq Anything more to do here? I think this pull request is ready, Neovim is still missing support for `hlget()`.

@shadmansaleh It looks like there are some merge conflicts that need to be resolved.

zeertzjq commented:
A lot of merge conflicts, not some :)

shadmansaleh commented:
Currently, I don't have the time or motivation to work on this. If anyone wants, can pick it up from here.

justinmk commented:
Do we actually need these? If they are addressed by `nvim_get_hl` and `nvim_set_hl` then I propose that we:

1. mark these patches as N/A
2. add a note in `vim_diff.txt`

Going forward, we should avoid adding new Vim functionality if it duplicates features we already have. But it's also important to (1) mark such patches as N/A (so they don't appear in the `vim-patch.sh -l` list), and (2) mention that this is intentional in `vim_diff.txt`.

vimpostor commented:
> Going forward, we should avoid adding new Vim functionality if it duplicates features we already have.

I think this is a ridiculous example for commenting this opinion, given that vim's `hlget()` existed years before `nvim_get_hl()`. In fact this very PR intending to port that function was submitted a whole 1.5 years **before** neovim even had the [`nvim_get_hl()` replacement](https://github.com/neovim/neovim/pull/22693).

In fact I would argue for the polar opposite: If there is no technical reason why vim API wouldn't work with neovim, then there is no harm in porting over the patches so that the same API works in neovim.

I would go even further and say that there is no reason to recreate equivalent functionality with completely different API like it was done in #22693. This just unnecessarily splits apart the vim communities even more, where there is absolutely no technical reason (`nvim_get_hl()` is written in C, so there was literally not a single reason to create different API for this).

Let's also not ignore that this PR was stalled for a year despite all review comments having been addressed.

justinmk commented:
> I would go even further and say that there is no reason to recreate equivalent functionality with completely different API like it was done in #22693.

I agree. 

But the sequence matters, even if it happened by accident. Somehow or other we got `nvim_set_hl`/`nvim_get_hl`, so now it doesn't make sense to also add redundant things, whatever those redundant things are, and irrespective of the timestamp on the Vim commit.

> This just unnecessarily splits apart the vim communities even more

There is a fork since Vim 8, that was made clear. At this point it only makes sense for us to merge upstream Vim things that reduce _maintenance costs_, there is no goal of trying to match APIs.

P.S. `nvim_set_hl`/`nvim_get_hl` have "namespace" support, so they are not merely a reinvention of `hlget()` and friends. But that's orthogonal to the general point.

zeertzjq commented:
> there is no technical reason why vim API wouldn't work with neovim

This is not possible, because Nvim has had highlight namespaces since 2020 :)

shadmansaleh commented:
Duplicating functionality is unnecessary. Neovim already has way too many ways of doing the same thing in many cases. Marking these N/A sounds all right to me.

If we want to keep vim compatibility, writing a shim around existing nvim_get_hl and nvim_set_hl makes more sense. It can probably be done in lua too. Don't necessarily need to bring in vims implementation.
